eval/topdown_test.go

```diff
@@ -564,46 +433,27 @@ func TestTopDownVarReferences(t *testing.T) {
 	data := loadSmallTestData()
 
 	for i, tc := range tests {
+		runTopDownTestCase(t, data, i, tc.note, tc.rules, tc.expected)
+	}
+}
 
-		rules := parseRules(tc.rules)
-		store := NewStorageFromJSONObject(data)
-		for _, rule := range rules {
-			store[string(rule.Name)] = rule
-		}
-
-		ctx := &TopDownContext{
-			Rule:     rules[0],
-			Store:    store,
-			Bindings: newHashMap(),
-		}
+func TestTopDownDisjunction(t *testing.T) {
 
-		switch e := tc.expected.(type) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"incr: query set", []string{"p[x] :- a[i] = x", "p[y] :- b[j] = y"}, `[1,2,3,4,"hello","goodbye"]`},
+		{"incr: query object", []string{"p[k] = v :- b[v] = k", "p[k] = v :- a[i] = v, g[k][j] = v"}, `{"b": 2, "c": 4, "hello": "v1", "goodbye": "v2", "a": 1}`},
+		{"incr: eval set", []string{"p[x] :- q[x]", "q[x] :- a[i] = x", "q[y] :- b[j] = y"}, `[1,2,3,4,"hello","goodbye"]`},
+		{"incr: eval object", []string{"p[k] = v :- q[k] = v", "q[k] = v :- b[v] = k", "q[k] = v :- a[i] = v, g[k][j] = v"}, `{"b": 2, "c": 4, "hello": "v1", "goodbye": "v2", "a": 1}`},
+	}
```

Comment:
It'd be good to have a test where we ensure we're passing variable bindings properly from one rule to the next.  Not sure about whether the test makes sense for the base documents that you have...
p[x]=y :- r[z]=y, q[x]=z 
q[x] = y :- s[x]=y
But here we're checking that the y from the p rule is not being bound when evaluating the q rule.